### DISCUSSION TITLE
What triggers "custom badge invalid"?

### DISCUSSION THREAD

**Discussion by andypiper** [badges/shields#5710] [Q&A]
Hello! Thanks for the Shields service, long time user, first time / recent developer of custom badges.

I've built a very simple JSON server to power some custom badges -> https://twbadges.glitch.me/

As of today, I'm seeing inconsistent behaviour for the badges. Sometimes, some of them, show up with a grey badge stating "custom badge | invalid".

I'm trying to track down the best way to debug this. So far as I can see, when I reload the page, the number of calls from shields.io back to the server is itself inconsistent, resulting in the "dropped" badges. For example, when I reloaded the page just now, I saw 5 hits on the JSON backend, corresponding to 5 correctly-rendered badges; and two that show as invalid.

So, my assumption is that there's a bottleneck on the Shields side itself. Clearly this particular page is a showcase / demo rather than the way I expect the badges themselves to be used; I'd imagine one, two, at most three, might be added to a single README file; so I don't expect the kind of constant reloading I'm causing as I test the front-end web page.

I suppose I could build an "API version picker" for a developer to select the badge they want and render them individually on-demand, which might be better practice / better performance / a way to be a nicer citizen to Shields?

Any thoughts on the behaviour I'm seeing would be appreciated :-) thank you!

**Comment by chris48s** [comment]
Hmm. I'm seeing the same problem as you, but I don't know what's causing it.
If I hit https://twbadges.glitch.me/badges/adsv7 a bunch of times, I get a valid response. If I hit https://img.shields.io/endpoint?url=https%3A%2F%2Ftwbadges.glitch.me%2Fbadges%2Fv2 a bunch of times, sometimes I get an invalid. It does appear to be an issue with us rather than the JSON endpoint.
If you're not seeing a hit on the JSON endpoint for the ones that don't render, I wonder if we're getting rate-limited by glitch? I reckon if one of our nodes has been rate-limited but not all of them we would see that kind of inconsistent behaviour. Do you know anything about rate limits that glitch.me might apply?

**Reply by andypiper** [reply]
I do not know about their rate limits, but that’s perfectly possible! Good thought. I’ll go dig to see what I can find.

**Reply by chris48s** [reply]
I tried this again today and I can't reproduce it now.

One other thing I've realised is that your badges are talking to the twitter API, and twitter was having some issues yesterday (and still today by the looks of it) https://api.twitterstat.us/#past-incidents so I wonder if that could have also been related?

**Reply by andypiper** [reply]
Thanks for following up. The badges do not talk to Twitter directly, they are based on a static JSON file that is part of the server, so this was not the issue, but the timings would indeed seem to have lined up.

**Reply by andypiper** [reply]
I'm now thinking more in terms of a memory leak on the Glitch app side as I noticed the whole thing went down at a certain point and I had to run some commands manually to get it working again, so that's also a possible angle here. Thanks again for the help!
